compiler/rustc_lint/src/types/improper_ctypes.rs

```diff
@@ -291,57 +346,133 @@ impl VisitorState {
     const RETURN_TY_IN_FNPTR: Self =
         Self::from_bits(Self::FUNC.bits() | Self::THEORETICAL.bits() | Self::FN_RETURN.bits())
             .unwrap();
+}
+
+impl EphemeralStateFlags {
+    /// modify self to change the ephemeral part of the flags
+    fn from_outer_ty<'tcx>(ty: Ty<'tcx>) -> Self {
+        match ty.kind() {
+            ty::FnPtr(..) => Self::NO_OUTER_TY | Self::NOOUT_FNPTR,
+            k @ (ty::RawPtr(..) | ty::Ref(..)) => {
+                let mut ret = Self::IN_PTR;
+                if ty.is_mutable_ptr() {
+                    ret |= Self::PTR_MUT;
+                }
+                if matches!(k, ty::RawPtr(..)) {
+                    ret |= Self::PTR_RAW;
+                }
+                ret
+            }
+            ty::Adt(..) => {
+                if ty.boxed_ty().is_some() {
+                    Self::IN_PTR | Self::PTR_MUT
+                } else {
+                    Self::IN_ADT | Self::MEMORY_INLINED
+                }
+            }
+            ty::Tuple(..) | ty::Array(..) | ty::Slice(_) => Self::MEMORY_INLINED,
+            k @ _ => bug!("Unexpected outer type {:?} of kind {:?}", ty, k),
+        }
+    }
+}
+
+impl VisitorState {
+    /// From an existing state, compute the state of any subtype of the current type.
+    fn get_next<'tcx>(&self, current_ty: Ty<'tcx>) -> Self {
+        debug_assert!(!matches!(current_ty.kind(), ty::FnPtr(..)));
+        Self {
+            persistent_flags: self.persistent_flags,
+            ephemeral_flags: EphemeralStateFlags::from_outer_ty(current_ty),
+        }
+    }
+    fn get_next_in_fnptr<'tcx>(&self, current_ty: Ty<'tcx>, is_ret: bool) -> Self {
```

Comment:
Optional nit: adding an enum rather than using a bool may come in handy in a few places

```rust
enum FnPos {
    Arg,
    Ret,
}
```